File: Python/Maximal Rectangle.py
# Given a 2D binary matrix filled with 0's and 1's, find the largest rectangle containing all ones and return its area.

class Solution:
    # @param matrix, a list of lists of 1 length string
    # @return an integer
    def maximalRectangle(self, matrix):
		row = len(matrix)
		if row == 0:
			return 0
		res = 0
		col = len(matrix[0])
		ones = [[0 for i in xrange(col)] for j in xrange(row)]
		for i in xrange(row):
			for j in xrange(col):
				if matrix[i][j] != "0":
					if j == 0:
						ones[i][j] = 1
					else:
						ones[i][j] = ones[i][j-1] + 1
		for i in xrange(row):
			for j in xrange(col):
				if ones[i][j] != 0:
					h = i - 1
					temp = ones[i][j]
					mini = ones[i][j]
					while h >= 0:
						if ones[h][j] == 0:
							break
						else:
							if ones[h][j] < mini:
								mini = ones[h][j]
							if temp < mini*(i-h+1):
								temp = mini*(i-h+1)
							h = h - 1
					if res < temp:
						res = temp
		return res	

File: Java/Maximal Rectangle.java
class Maximal Rectangle{
	public int maximalRectangle(char[][] matrix) {
        if(matrix == null || matrix.length == 0)
        	return 0;
        int rowLen = matrix.length;
        int colLen = matrix[0].length;
        int[][] ones = new int[rowLen][colLen];
        for(int i=0;i<rowLen;i++){
        	for(int j=colLen-1;j>=0;j--){
        		if(j == colLen-1)
        			ones[i][j] = matrix[i][j] - '0';
        		else{
        			if(matrix[i][j] == '1')
        				ones[i][j] = 1+ones[i][j+1];
        			else
        				ones[i][j] = 0;
        		}
        	}
        }
        int max = 0;
        int currArea = 0;
        for(int i=0;i<rowLen;i++){
        	for(int j=0;j<colLen;j++){
        		currArea = ones[i][j];
        		int minWidth = ones[i][j];
        		int height = i+1;
        		while(height < rowLen){
        			minWidth = Math.min(minWidth,ones[height][j]);
        			if(minWidth == 0)
        				break;
        			currArea = Math.max(currArea,minWidth*(height-i+1));
        			height++;
        		}
        		max = Math.max(max,currArea);
        	}
        }
        return max;
    }
} 